File: problems/arrays_and_hashing/cpp/1929_concatenation_of_array.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * @complexity
 * - Time: O(n), where n is the size of the input vector. We iterate
 * over the vector once to fill the result.
 * - Space: O(n), where n is the size of the input vector. We
 * allocate additional space for the concatenated result.
 */
class Solution {
public:
  vector<int> get_concatenation(vector<int> &nums) {
    // Return an empty vector if input is empty
    if (nums.empty()) {
      return {};
    }

    // Allocate space for result, which is twice the size of the input vector
    int n = nums.size();
    vector<int> result(2 * n);

    // Copy nums twice: once into the first half and again into the second half
    for (int i = 0; i < n; ++i) {
      result[i] = nums[i];     // Copy original to the first half
      result[i + n] = nums[i]; // Copy original to the second half
    }

    return result;
  }
};

File: problems/arrays_and_hashing/rust/1929_concatenation_of_array.rs
struct Solution;

impl Solution {
    pub fn get_concatenation(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut result: Vec<i32> = vec![0; n * 2];
        for index in 0..n {
            result[index] = nums[index];
            result[index + n] = nums[index];
        }
        result
    }
}

fn main() {
    let _sol = Solution;
}
